### PR TITLE
Add internal boundary element support for NCMesh

### PR DESCRIPTION
**Pull request by hughcars**
Currently if a mesh includes any internal boundary elements, these will be duplicated once a mesh is made NC and parallelized. This is a function of boundary elements being added by looping over elements, then looping over faces. Thus any face can be visited twice. This PR expands the `GetBoundaryClosure` method to also uncover hidden master boundary faces and mark these as essential.

I consider the previous behaviour of the NCMesh a bug, as it would result in the number of boundary elements changing as a mesh with an NCMesh was used to initialize a ParMesh, and any functionals on internal boundaries were doubled. 

To test this several additional tests have been added to `test_ncmesh.cpp`, primarily on the serial code as this demonstrated most of the issues, there are also parallel tests.

This PR is based off of the top of https://github.com/mfem/mfem/pull/3611 which will require merging first. This is due to issues as the number of tests in `test_ncmesh.cpp` has been growing rapidly, placing them in this sequence makes the eventual merging simpler.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3746](https://github.com/mfem/mfem/pull/3746) | @hughcars | @tzanio | @tzanio + @dylan-copeland + @psocratis | 11/14/23 | 1/2/24 | 1/4/24 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by hughcars** [comment]
Further testing of this on some larger parallel examples has uncovered an issue, that I am now able to recreate in a unit test, but which is causing a bit of a problem. 

Internal boundaries are used for two things: 1) computing functionals/applying Robin boundary conditions on the boundary, and 2) applying essential boundary conditions. 

For functionals/Robin conditions, the most appropriate location for the integrals is on the child faces, as within each face all bases variables are polynomial, thus the quadrature is accurate (integrating on the master would require complicated quadrature rules based on sub faces). To this end, during an NC refinement a boundary element is defined to be a conformal face, or a child face. 

For essential boundary conditions however, the true dofs associated with a face must be encountered. Child faces generally have few or even no true dofs associated to them (maybe one corner for instance). For P1 H1 elements, marking all slave dofs as essential, captures all the P1 master dofs too. So for P1, visiting the master face dofs, or the slave face dofs, and applying essential boundary conditions, results in the same relation between $P^T$ and $R$ that is tested for in `pfespace.cpp`. 

For P2 and P3 upwards however, there are face dofs within the master face that are not related to those of the child faces. These can only be found when visiting the master face, so when 
```
void FiniteElementSpace::GetEssentialVDofs(const Array<int> &bdr_attr_is_ess,
                                           Array<int> &ess_vdofs,
                                           int component) const
```
is called, each slave face must use the NCMesh member structures to get the master face index, then extract the dofs associated with that Face using `GetFaceDofs` and `GetFaceVDofs`. This works perfectly in serial, and for a large number of parallel meshes, there is however an edge case though. 

If the master face is a _ghost master face_, then `GetFaceGeometry`, a key substep, fails:
```
         if (Face < NumOfFaces) // local (non-ghost) face
         {
            return faces[Face]->GetGeometryType();
         }
         // ghost face
         const int nc_face_id = faces_info[Face].NCFace;

         MFEM_ASSERT(nc_face_id >= 0, "parent ghost faces are not supported");
         return faces[nc_faces_info[nc_face_id].MasterFace]->GetGeometryType();
 ```
In essence, there is not enough information being stored about ghost master faces to be able to constrain all the necessary dofs. 

This begs the question: Why was it decided not to populate this information? I'm on the verge of attempting to build this as part of the NC construction, but want to know if there's a good reason it didn't already exist, to avoid time wasting.

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: mesh/ncmesh.cpp [hunk 1]
```diff
@@ -2449,18 +2479,21 @@ void NCMesh::GetMeshComponents(Mesh &mesh) const
...
-   for (int i = 0; i < mesh.NumOfBdrElements; i++)
+   for (auto &elem : mesh.boundary)
    {
-      mesh.FreeElement(mesh.boundary[i]);
+      mesh.FreeElement(elem);
    }
    mesh.boundary.SetSize(0);
 
+   // Save off boundary face vertices to make boundary elements later.
+   std::map<int, mfem::Array<int>> unique_boundary_faces;
```
Comment (hughcars) [inline mesh/ncmesh.cpp:2495]: This section is the piece that adds in boundaries if they are not master faces.

**Comment by tzanio** [comment]
Can you please resolve the conflicts with `master`?

File: mesh/ncmesh.cpp [hunk 2]
```diff
@@ -5228,43 +5292,56 @@ void NCMesh::GetBoundaryClosure(const Array<int> &bdr_attr_is_ess,
                   bdr_edges.Append(nodes[enode].edge_index);
                }
             }
+
+            // If the face is a slave face, collect any non-ghost master face
```
Comment (hughcars) [inline mesh/ncmesh.cpp:5296]: This is the section which handles "hidden" boundary faces, namely a master boundary face coincident to a processor boundary. If all the slave faces neighbor to this face are on another rank, this face might not know it's dof need to be marked essential. The `bdr_faces` variable is used to account for this, collecting up such faces so that their dof can be collected and marked as essential vdofs.

File: mesh/ncmesh.cpp [hunk 3]
```diff
@@ -2651,21 +2702,22 @@ void NCMesh::OnMeshUpdated(Mesh *mesh)
       for (int j = 0; j < gi.nf; j++)
       {
          const int *fv = gi.faces[j];
-         Face* face = faces.Find(el.node[fv[0]], el.node[fv[1]],
-                                 el.node[fv[2]], el.node[fv[3]]);
-         MFEM_ASSERT(face, "face not found!");
+         int fid = faces.FindId(el.node[fv[0]], el.node[fv[1]],
```
Comment (dylan-copeland) [inline mesh/ncmesh.cpp:2705]: `const int fid`

File: mesh/ncmesh.cpp (hunk 2 above)
Comment (dylan-copeland) [reply mesh/ncmesh.cpp:5296]: Multiple comments in this file mention "any masters." Would it be more accurate to say "its master", since each slave handles its one master? Using the plural suggests multiple masters, which I suppose could mean masters at higher levels in the hierarchy, but that is not the case here.

File: mesh/pncmesh.cpp [hunk 4]
```diff
@@ -577,41 +577,81 @@ void ParNCMesh::CalcFaceOrientations()
    face_orient.SetSize(NFaces);
    face_orient = 0;
 
-   for (auto face = faces.begin(); face != faces.end(); ++face)
+   for (auto face : faces)
```
Comment (dylan-copeland) [inline mesh/pncmesh.cpp:580]: `for (const auto &face : faces)`

File: mesh/pncmesh.cpp [hunk 5]
```diff
@@ -577,41 +577,81 @@ void ParNCMesh::CalcFaceOrientations()
...
+         for (int j = mf.slaves_begin; j < mf.slaves_end; j++)
+         {
+            const auto &sf = GetFaceList().slaves[j];
+            if (sf.index < 0)
+            {
+               // Edge-face constraint. Skip this edge.
+               continue;
+            }
+            Face *face = GetFace(elements[sf.element], sf.local);
+            if (face && face->Boundary() && bdr_attr_is_ess[face->attribute - 1])
```
Comment (dylan-copeland) [inline mesh/pncmesh.cpp:617]: Out of curiosity, and maybe to prompt more detailed documentation, in what case would `face` be null? Is it the case of this slave being a master in the hierarchy, so it is not in the mesh?

File: fem/fespace.cpp [hunk 6]
```diff
@@ -944,6 +971,15 @@ int FiniteElementSpace::GetEntityDofs(int entity, int index, Array<int> &dofs,
    }
 }
 
+int FiniteElementSpace::GetEntityVDofs(int entity, int index, Array<int> &dofs,
+                                       Geometry::Type master_geom,
+                                       int variant) const
+{
+   int n = GetEntityDofs(entity, index, dofs, master_geom, variant);
```
Comment (dylan-copeland) [inline fem/fespace.cpp:978]: `const int n`

File: mesh/ncmesh.cpp (hunk 2 above)
Comment (hughcars) [reply mesh/ncmesh.cpp:5296]: The use of any was being coloured by my thinking of faces as three buckets (slave, master, conformal) during this work rather than their relation between individual faces. I think you're correct that its is more precise here, as there's a specific pair of faces and specific relation between them.

File: fem/gridfunc.hpp [hunk 7]
```diff
@@ -586,6 +586,10 @@ class GridFunction : public Vector
       return ComputeLpError(infinity(), exsol, NULL, NULL, irs);
    }
 
+   virtual double ComputeL1Error(Coefficient *exsol[],
```
[the remark this answers is not in the document]
Comment (hughcars) [reply fem/gridfunc.hpp:589]: This is following the pattern for the passing of `ComputeL2Error`, where there's an alternative method for this array of pointers of coefficients. The `ComputeW11Error` method is used in the pgrid function version, and what I noticed is that because that function isn't marked as override, there doesn't exist a matching `gridfunc.hpp` original version of the method (this is also the textbook example of why `override` should be used). Matching the usage on the array of pointers `ComputeL2Error` call in `pgridfunc.hpp`, I add the serial call in grid function which would be overriden.

File: fem/pfespace.cpp [hunk 8]
```diff
@@ -2161,6 +2169,11 @@ void NeighborRowMessage::Decode(int rank)
             int fo = pncmesh->GetFaceOrientation(id.index);
             ind = fec->DofOrderForOrientation(geom, fo);
          }
+         // P2 tri faces have dofs that must be processed in pairs, as the doftransformation
```
[the remark this answers is not in the document]
Comment (hughcars) [reply fem/pfespace.cpp:2172]: Second order polynomials.

File: mesh/pncmesh.cpp (hunk 5 above)
Comment (hughcars) [reply mesh/pncmesh.cpp:617]: I'm not aware of a case where it would be. As the local face numbers on the slaves should always be valid. I was using the `GetFace` api that returned a pointer by default however, and whenever accessing a pointer I put the check out of cleanliness. In this case however I can force the survival of the temporary and bind to `const&` and not worry given the extra information I have on the validity of the arguments.

File: fem/gridfunc.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (dylan-copeland) [reply fem/gridfunc.hpp:588]: Thanks for the explanation. I see now that this new function adds what was missing to follow the pattern used by L2. I wasn't aware that these norms were computed differently, for slightly different function signatures, and I am curious about the reasoning behind that. After briefly reviewing the existing functions `ComputeW11Error` and `ComputeLpError` for the L1 case, it appears that they do the same computation with different implementation (e.g. one uses GetValues whereas the other computes values via shape functions). I wonder whether they are really the same up to round-off, and if so, why do we use different functions that compute the same thing? Maybe someone familiar with the history of these functions can comment. Hopefully some clarification can lead to more documentation, since I am a little confused still. It would be nice to have documentation on these functions to say what they compute and whether they are the same as other functions with different signatures.

File: tests/unit/mesh/test_ncmesh.cpp [hunk 9]
```diff
@@ -887,147 +634,2169 @@ TEST_CASE("PNQ2PureTetHexPri",  "[Parallel], [NCMesh]")
...
+         e_i = 0.0;
+         e_i(i) = 1.0;
+         R->MultTranspose(e_i, Rrow);
+         for (int j = 0; j < P->Width(); j++)
+         {
+            e_j = 0.0;
+            e_j(j) = 1.0;
+            P->Mult(e_j, Pcol);
+
+            CHECK(Rrow * Pcol == (i == j ? 1.0 : 0.0));
```
Comment (dylan-copeland) [inline tests/unit/mesh/test_ncmesh.cpp:2729]: Update a bool and put one check after the loops? Also below at line 2744.

**Comment by dylan-copeland** [comment]
CI is failing just for Windows. Maybe it needs `#include <array>`?
D:\a\mfem\mfem\mfem\tests\unit\mesh\mesh_test_utils.cpp(642,23): error C2027: use of undefined type 'std::array<double,2>'

**Review by dylan-copeland — APPROVED** [review]
Thanks @hughcars! The documentation is very helpful. I am still curious about `ComputeL1Error` and would like to have some documentation on the differences between versions of norms.

File: tests/unit/mesh/mesh_test_utils.hpp [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+ * condition.
+ */
+int CheckPoisson(Mesh &mesh, int order, int disabled_boundary_attribute = -1);
+
+/**
+ * @brief Helper for counting the number of essential degrees of freedom within
+ * a mesh.
+ *
+ * @tparam FECollection FiniteElementCollection to define the space.
+ * @tparam TDOF Whether or not to counter true degrees of freedom (TDOF) or
```
Comment (psocratis) [inline tests/unit/mesh/mesh_test_utils.hpp:45]: ... to count ...

File: tests/unit/mesh/mesh_test_utils.hpp [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+ * @param attribute The attribute of the boundary to count the essential DOFs
+ * on.
+ * @return int The number of essential DOFs.
+ */
+template <typename FECollection, bool TDOF = true>
+int CountEssentialDof(Mesh &mesh, int order, int attribute);
+
+/**
+ * @brief Build a mesh with a central tetrahedron surrounded by four
+ * tetrahedron, one sharing each face with the central tetrahedron.
```
Comment (psocratis) [inline tests/unit/mesh/mesh_test_utils.hpp:58]: tetrahedrons/tetrahedra

File: tests/unit/mesh/mesh_test_utils.hpp [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+ */
+template <typename FECollection, bool TDOF = true>
+int CountEssentialDof(ParMesh &mesh, int order, int attribute);
+
+/**
+ * @brief Helper for counting the number of essential degrees of freedom within
+ * a parallel mesh, and summing over all processors.
+ *
+ * @tparam FECollection FiniteElementCollection to define the space.
+ * @tparam TDOF Whether or not to counter true degrees of freedom (TDOF) or
```
Comment (psocratis) [inline tests/unit/mesh/mesh_test_utils.hpp:192]: ... count ...

File: tests/unit/mesh/mesh_test_utils.hpp [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+ * @return true The identity holds
+ * @return false The identity does not hold
+ */
+bool CheckRPIdentity(const ParFiniteElementSpace& pfespace);
+
+#endif
+
+} // namespace mfem
+
+#endif // MFEM_MESH_TEST_UTILS
```
Comment (psocratis) [inline tests/unit/mesh/mesh_test_utils.hpp:215]: add empty line at the end

**Review by psocratis — APPROVED** [review]

File: tests/unit/mesh/mesh_test_utils.hpp (hunk 10 above)
Comment (hughcars) [reply tests/unit/mesh/mesh_test_utils.hpp]: Fixed in https://github.com/mfem/mfem/pull/3746/commits/4fe30263e4beb521564a5a2561501e8bbd67900c

File: tests/unit/mesh/mesh_test_utils.hpp (hunk 11 above)
Comment (hughcars) [reply tests/unit/mesh/mesh_test_utils.hpp]: Fixed in https://github.com/mfem/mfem/pull/3746/commits/4fe30263e4beb521564a5a2561501e8bbd67900c

File: tests/unit/mesh/mesh_test_utils.hpp (hunk 12 above)
Comment (hughcars) [reply tests/unit/mesh/mesh_test_utils.hpp]: Fixed in https://github.com/mfem/mfem/pull/3746/commits/4fe30263e4beb521564a5a2561501e8bbd67900c

File: tests/unit/mesh/mesh_test_utils.hpp (hunk 13 above)
Comment (hughcars) [reply tests/unit/mesh/mesh_test_utils.hpp]: Fixed in https://github.com/mfem/mfem/pull/3746/commits/4fe30263e4beb521564a5a2561501e8bbd67900c

**Comment by hughcars** [comment]
@dylan-copeland @psocratis I think this is ready for `next`?

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
There was an autotest error last night in this run: "mpirun -np 4 ex6p -m ../data/inline-segment.mesh -o 1 -md 200 -no-vis".

@hughcars, can youn take a look if it is due to this PR?

**Comment by hughcars** [comment]
> There was an autotest error last night in this run: "mpirun -np 4 ex6p -m ../data/inline-segment.mesh -o 1 -md 200 -no-vis".
> 
> @hughcars, can youn take a look if it is due to this PR?

The fault was mine, didn't to treat points in the boundary face discovery correctly. Simple fix in https://github.com/mfem/mfem/pull/3746/commits/5736f3832b3a929695c386396a875fd812bc9ee9

**Comment by tzanio** [comment]
Re-merged in `next` for testing...
